src/transformers/modeling_utils.py

```diff
@@ -833,21 +852,40 @@ def _generate_no_beam_search(
                 next_token = torch.argmax(next_token_logits, dim=-1)
 
             # update generations and finished sentences
-            tokens_to_add = next_token * unfinished_sents + pad_token_id * (1 - unfinished_sents)
+            if eos_token_ids is not None:
+                # if pad_token_id is undefined, use first eos_token_id since sentence will be filled with -1 anyways
+                tokens_to_add = next_token * unfinished_sents + (pad_token_id) * (1 - unfinished_sents)
+            else:
+                tokens_to_add = next_token
+
             input_ids = torch.cat([input_ids, tokens_to_add.unsqueeze(-1)], dim=-1)
-            for eos_token_id in eos_token_ids:
-                unfinished_sents.mul_(tokens_to_add.ne(eos_token_id).long())
+
+            if eos_token_ids is not None:
+                for eos_token_id in eos_token_ids:
+                    eos_in_sents = tokens_to_add == eos_token_id
+                    # batches_len is updated only the first time eos_token is in the sentences (unfinished_sents == 1 and eos_in_sents == 1)
+                    batches_len.masked_fill_(unfinished_sents.mul(eos_in_sents.long()).bool(), cur_len + 1)
+                    # unfinished_sents is set to zero if eos in sentence
```

Comment:
love this comment, but I don't totally understand the multiplication. Sorry if this is dumb!

Isn't the code multiplying is_the_sentence_already_unfinished * not(does the sentence have eos)?
In the case where the sentence wasnt finished, and there is now an EOS in the sentence, do we get 0 (leave leave it unfinished)

Reply:
Dissected the equation to make it more understandable and improved comment!

Reply:
So the idea here is that if the sentence i is unfinished: unfinished_sents[i] == 1 and there is and the next token is an eos token (== 1) only then should the batches_len( now renamed to sent_lengths) be filled with the current length. Hope the new code is clearer to understand!